### PR TITLE
More value inference for dbutils.notebook.run

### PR DESCRIPTION
**Pull request by ericvergnaud**
add tests for value inference scenarios
support multiple values inference when `dbutils.notebook.run` is called in a loop

### Linked issues
Progress #1205

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests
- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/graph.py [hunk 1]
```diff
@@ -185,11 +185,16 @@ def _process_node(self, base_node: NodeBase):
...
+            paths = base_node.get_notebook_paths()
+            if None in paths:
+                yield DependencyProblem(
+                    'dependency-too-complex',
+                    f"Can't check dependency from {base_node.node.as_string()} because the expression is too complex",
+                )
             else:
-                yield from self.register_notebook(Path(strpath))
+                for path in paths:
+                    assert isinstance(path, str)
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/graph.py:196]: Why this assert?

File: src/databricks/labs/ucx/source_code/linters/imports.py [hunk 2]
```diff
@@ -83,12 +84,23 @@ class NotebookRunCall(NodeBase):
...
+        except InferenceError:
+            logger.debug(f"Can't infer value(s) of {node.as_string()}")
+            return [None]
+
+    @classmethod
+    def _get_notebook_paths(cls, nodes: Iterable[NodeNG]) -> list[str | None]:
+        paths: list[str | None] = []
+        for node in nodes:
+            if isinstance(node, Const):
+                paths.append(node.as_string().strip("'").strip("'"))
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/linters/imports.py:100]: why stripping the same character twice?

File: src/databricks/labs/ucx/source_code/graph.py [hunk 3]
```diff
@@ -185,11 +185,16 @@ def _process_node(self, base_node: NodeBase):
         if isinstance(base_node, SysPathChange):
             self._mutate_path_lookup(base_node)
         if isinstance(base_node, NotebookRunCall):
-            strpath = base_node.get_notebook_path()
-            if strpath is None:
-                yield DependencyProblem('dependency-not-constant', "Can't check dependency not provided as a constant")
+            paths = base_node.get_notebook_paths()
+            if None in paths:
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/graph.py:189]: We don't want to try to register the not `None` paths?

File: src/databricks/labs/ucx/source_code/graph.py [hunk 4]
```diff
@@ -185,11 +185,16 @@ def _process_node(self, base_node: NodeBase):
...
+                yield DependencyProblem(
+                    'dependency-too-complex',
+                    f"Can't check dependency from {base_node.node.as_string()} because the expression is too complex",
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/graph.py:192]: I would refrain from using "too complex" it is subjective, what makes something complex?

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
Added some comments

File: src/databricks/labs/ucx/source_code/graph.py (hunk 1 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:196]: to avoid a linter warning for str | None

File: src/databricks/labs/ucx/source_code/linters/imports.py (hunk 2 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/imports.py:100]: Typo! Thanks

File: src/databricks/labs/ucx/source_code/graph.py (hunk 3 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:189]: good point

File: src/databricks/labs/ucx/source_code/graph.py (hunk 4 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py]: changed

File: src/databricks/labs/ucx/source_code/notebooks/loaders.py [hunk 5]
```diff
@@ -55,17 +55,20 @@ def load_dependency(self, path_lookup: PathLookup, dependency: Dependency) -> So
...
             logger.warning(f"Could not detect language for {absolute_path}")
             return None
         return Notebook.parse(absolute_path, content, language)
 
     @staticmethod
-    def _detect_language(content: str):
-        for language in CellLanguage:
-            if content.startswith(language.file_magic_header):
-                return language.language
+    def detect_language(path: Path, content: str):
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebooks/loaders.py:65]: for testing

File: src/databricks/labs/ucx/source_code/linters/imports.py [hunk 6]
```diff
@@ -104,19 +116,20 @@ def lint(self, code: str) -> Iterable[Advice]:
...
                 node.lineno,
                 node.col_offset,
                 node.end_lineno or 0,
                 node.end_col_offset or 0,
             )
         return Advisory(
-            'dbutils-notebook-run-dynamic',
-            "Path for 'dbutils.notebook.run' is not a constant and requires adjusting the notebook path",
+            'dbutils-notebook-run-literal',
+            "Call to 'dbutils.notebook.run' will be migrated automatically",
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/imports.py:132]: Sowaht should be the message (the above was existing)

File: src/databricks/labs/ucx/source_code/linters/imports.py [hunk 7]
```diff
@@ -83,12 +84,23 @@ class NotebookRunCall(NodeBase):
     def __init__(self, node: Call):
         super().__init__(node)
 
-    def get_notebook_path(self) -> str | None:
-        node = DbutilsLinter.get_dbutils_notebook_run_path_arg(cast(Call, self.node))
-        inferred = next(node.infer(), None)
-        if isinstance(inferred, Const):
-            return inferred.value.strip().strip("'").strip('"')
-        return None
+    def get_notebook_paths(self) -> list[str | None]:
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/imports.py:87]: astroid is clever enough to return multiple nodes in a scenario such as:
```
paths = [ "p1", "p2"]
for path in paths:
    dbutils.notebook>run(path)
```

File: src/databricks/labs/ucx/source_code/notebooks/loaders.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/ucx/source_code/notebooks/loaders.py:65]: it's a slippery slope to expose methods public just for testing without a significant need. This case doesn't justify this need and could be tested through other public methods.

File: src/databricks/labs/ucx/source_code/graph.py (hunk 1 above)
Comment (nfx) [reply src/databricks/labs/ucx/source_code/graph.py]: It'll crash in production with AssertionError. Can we avoid None values here?

Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py]: it won't crash since None is already tested above. It's a linter limitation

File: tests/unit/source_code/linters/test_python_ast.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+name = "xyz" + "-" + "abc"
+dbutils.notebook.run(name)
+""",
+            ["xyz-abc"],
+        ),
+        (
+            """
+names = ["abc", "xyz"]
+for name in names:
+    dbutils.notebook.run(name)
```
Comment (nfx) [inline tests/unit/source_code/linters/test_python_ast.py:113]: Can you also add few more tests:

x = dbutils.notebook 
x.run("foo")

And 

y = dbutils
y.notebook.run("foo")

And

z = dbutils.notebook.run
z("foo")

And False positive test: 

x = Some()
x.run("foo")

File: tests/unit/source_code/linters/test_python_ast.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    nodes = set()
+    count = 0
+    tree = Tree.parse("o.m1().m2().m3()")
+    for node in tree.walk():
+        nodes.add(node)
+        count += 1
+    assert len(nodes) == count
+
+
+@pytest.mark.parametrize(
```
Comment (nfx) [inline tests/unit/source_code/linters/test_python_ast.py:92]: This test should be in imports.py

File: tests/unit/source_code/notebooks/test_loader.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
+from pathlib import Path
+
+from databricks.labs.ucx.source_code.notebooks.loaders import NotebookLoader
+from databricks.sdk.service.workspace import Language
+
+
+def test_detects_language():
+    assert NotebookLoader.detect_language(Path("hi.py"), "stuff") == Language.PYTHON
```
Comment (nfx) [inline tests/unit/source_code/notebooks/test_loader.py:8]: Test through existing public API, don't add new public apis without significant reason

File: src/databricks/labs/ucx/source_code/notebooks/loaders.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebooks/loaders.py:65]: there's nothing dangerous about this method, so not sure why this one is slippery ? or maybe we should allow access to private methods in unit testing, such that we can actually write unit tests rather than slower and complex end-to-end tests ?

File: tests/unit/source_code/samples/run_notebooks_with_fstring.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
+# Databricks notebook source
+import datetime
+# Updated List of Notebooks
+path = './leaf2.py'
+notebooks_list = [
```
Comment (nfx) [inline tests/unit/source_code/samples/run_notebooks_with_fstring.py:5]: Please also add some functional tests with edge cases like resolved constant not pointing to an existing notebook

**Review by nfx — CHANGES_REQUESTED** [review]

File: tests/unit/source_code/linters/test_python_ast.py (hunk 8 above)
Comment (ericvergnaud) [reply tests/unit/source_code/linters/test_python_ast.py:113]: they already exist, see test_python_imports

File: src/databricks/labs/ucx/source_code/linters/imports.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/ucx/source_code/linters/imports.py:85]: Oh, interesting. Please add it as a code comment, so that the next time reading this code won't catch by surprise

File: tests/unit/source_code/linters/test_python_ast.py (hunk 8 above)
Comment (ericvergnaud) [reply tests/unit/source_code/linters/test_python_ast.py:113]: Can we do this in a separate PR ? It's a completely different scenario, only currently supported for `os.path.abspath`.

File: tests/unit/source_code/samples/run_notebooks_with_fstring.py (hunk 11 above)
Comment (ericvergnaud) [reply tests/unit/source_code/samples/run_notebooks_with_fstring.py:5]: Functional tests don't currently build dependency graphs, do we want to change that ? (we already have tests that point to missing notebooks)

File: tests/unit/source_code/linters/test_python_ast.py (hunk 9 above)
Comment (ericvergnaud) [reply tests/unit/source_code/linters/test_python_ast.py:92]: Done

File: tests/unit/source_code/notebooks/test_loader.py (hunk 10 above)
Comment (ericvergnaud) [reply tests/unit/source_code/notebooks/test_loader.py:8]: Done using a test class

File: src/databricks/labs/ucx/source_code/linters/imports.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/imports.py:124]: Should we drop this advice altogether ?

File: tests/unit/source_code/samples/run_notebooks_with_fstring.py (hunk 11 above)
Comment (nfx) [reply tests/unit/source_code/samples/run_notebooks_with_fstring.py:5]: > But functional tests don't currently build dependency graphs, do we want to change that ?

yep, please add that capability :) perhaps in a separate PR then

File: tests/unit/source_code/linters/test_python_ast.py (hunk 8 above)
Comment (nfx) [reply tests/unit/source_code/linters/test_python_ast.py]: of course. just not to forget: https://github.com/databrickslabs/ucx/issues/1864

File: src/databricks/labs/ucx/source_code/notebooks/loaders.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/ucx/source_code/notebooks/loaders.py]: slippery slope is that if we allow it for trivial cases, then inexperienced devs would expose inner workings of classes as Public API, resulting in a more fragile system. This codebase was in that state 6 months ago and we're not going back there.

File: src/databricks/labs/ucx/source_code/linters/imports.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/ucx/source_code/linters/imports.py:129]: sure

File: src/databricks/labs/ucx/source_code/linters/imports.py [hunk 12]
```diff
@@ -81,12 +82,28 @@ class NotebookRunCall(NodeBase):
...
+            return self._get_notebook_paths(node.infer())
+        except InferenceError:
+            logger.debug(f"Can't infer value(s) of {node.as_string()}")
+            return [None]
+
+    @classmethod
+    def _get_notebook_paths(cls, nodes: Iterable[NodeNG]) -> list[str | None]:
+        paths: list[str | None] = []
+        for node in nodes:
+            if isinstance(node, Const):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/imports.py:102]: log a debug if it's not a constant and continue loop

File: src/databricks/labs/ucx/source_code/linters/imports.py [hunk 13]
```diff
@@ -81,12 +82,28 @@ class NotebookRunCall(NodeBase):
...
+        """
+        node = DbutilsLinter.get_dbutils_notebook_run_path_arg(self.node)
+        try:
+            return self._get_notebook_paths(node.infer())
+        except InferenceError:
+            logger.debug(f"Can't infer value(s) of {node.as_string()}")
+            return [None]
+
+    @classmethod
+    def _get_notebook_paths(cls, nodes: Iterable[NodeNG]) -> list[str | None]:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/imports.py:99]:
```suggestion
    def _get_notebook_paths(cls, nodes: Iterable[NodeNG]) -> list[str]:
```
let's avoid nullability

File: tests/unit/source_code/linters/test_python_imports.py [hunk 14]
```diff
@@ -216,22 +147,27 @@ def test_args_count(code, expected):
...
+        (
+            """
+name = "xyz" + "-" + "abc"
+dbutils.notebook.run(name)
+""",
+            ["xyz-abc"],
+        ),
+        (
+            """
+names = ["abc", "xyz"]
```
Comment (nfx) [inline tests/unit/source_code/linters/test_python_imports.py:161]: can we already do 

```
def foo(): return "bar"
name = foo()
dbutils.notebook.run(name)
```

or does it require building a small type-aware interpreter?

**Review by nfx — CHANGES_REQUESTED** [review]

File: tests/unit/source_code/linters/test_python_imports.py (hunk 14 above)
Comment (ericvergnaud) [reply tests/unit/source_code/linters/test_python_imports.py:161]: we can! added corresponding test in `test_infers_dbutils_notebook_run_dynamic_value`

File: src/databricks/labs/ucx/source_code/linters/imports.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/imports.py]: done

File: src/databricks/labs/ucx/source_code/linters/imports.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/imports.py]: done

File: src/databricks/labs/ucx/source_code/notebooks/loaders.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebooks/loaders.py]: noted

File: src/databricks/labs/ucx/source_code/linters/imports.py (hunk 12 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/imports.py:103]: done

File: src/databricks/labs/ucx/source_code/linters/imports.py (hunk 13 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/imports.py]: done

File: tests/unit/source_code/linters/test_python_imports.py (hunk 14 above)
Comment (nfx) [reply tests/unit/source_code/linters/test_python_imports.py:159]: Nice!

**Review by nfx — APPROVED** [review]
Lgtm

**Comment by nfx** [comment]
@ericvergnaud make fmt fails

**Comment by ericvergnaud** [comment]
@nfx blocked by code review requests, not why I can't dismiss them
